File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TeleOpModes/FlywheelTuner.java
package org.firstinspires.ftc.teamcode.TeleOpModes;

import com.acmerobotics.dashboard.config.Config;
import com.qualcomm.robotcore.eventloop.opmode.TeleOp;

import org.firstinspires.ftc.teamcode.Subsystems.FlywheelSubsystem;
import org.firstinspires.ftc.teamcode.Subsystems.FlywheelSystem;

import dev.nextftc.core.components.BindingsComponent;
import dev.nextftc.core.components.SubsystemComponent;
import dev.nextftc.ftc.ActiveOpMode;
import dev.nextftc.ftc.Gamepads;
import dev.nextftc.ftc.NextFTCOpMode;
import dev.nextftc.ftc.components.BulkReadComponent;
import dev.nextftc.hardware.controllable.MotorGroup;
import dev.nextftc.hardware.impl.MotorEx;
@Config
@TeleOp(name="Flywheel Tuner")
public class FlywheelTuner extends NextFTCOpMode {
    public FlywheelTuner() {
        addComponents(
                new SubsystemComponent(FlywheelSubsystem.INSTANCE),
                BindingsComponent.INSTANCE,
                BulkReadComponent.INSTANCE
        );
    }

    public static double velocity = 0;

    @Override
    public void onUpdate() {
        FlywheelSubsystem.INSTANCE.calibrateFlywheelVelocity(velocity);
        telemetry.addData("vel", FlywheelSubsystem.INSTANCE.getvel());
        telemetry.addData("targetvel", velocity);
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Resources/FlywheelTuner.kt
package org.firstinspires.ftc.teamcode.Resources

import com.acmerobotics.dashboard.config.Config
import dev.nextftc.control.KineticState
import dev.nextftc.control.builder.controlSystem
import dev.nextftc.control.feedback.PIDCoefficients
import dev.nextftc.control.feedforward.BasicFeedforwardParameters
import dev.nextftc.core.components.BindingsComponent
import dev.nextftc.ftc.NextFTCOpMode
import dev.nextftc.ftc.components.BulkReadComponent
import dev.nextftc.hardware.controllable.MotorGroup
import dev.nextftc.hardware.impl.MotorEx

@Config
class FlywheelTuner: NextFTCOpMode() {
    init {
        addComponents(
            BindingsComponent,
            BulkReadComponent
        )
    }

    companion object Constants {
        @JvmField
        val pidCoefficients = PIDCoefficients(0.0,0.0, 0.0)
        @JvmField
        val ffCoefficients = BasicFeedforwardParameters(0.0,0.0,0.0)
        @JvmField
        val vel = 0.0;
    }

    private val fwl = MotorEx("fwl").reversed()
    private val fwr = MotorEx("fwr")
    private val fwm = MotorGroup(fwr, fwl)

    val flywheelControl = controlSystem {
        velPid(pidCoefficients)
        basicFF(ffCoefficients)
    }

    override fun onUpdate() {
        flywheelControl.goal = KineticState(0.0, vel);
        fwm.power = flywheelControl.calculate(fwm.state)
    }
}